### ISSUE TITLE
Allow the  plugin DSL to expand properties as part of the version

### ISSUE THREAD

**Issue by JLLeitschuh** [gradle/gradle#9830]

**Comment by rgoldberg** [comment]
From what I’ve read elsewhere, Kotlin `plugins` accepts ad hoc code, but properties values are not stored in variables that are in scope in `plugins`.

Will the resolution for this issue introduce property variables into this scope, and possibly others? (Would that even be possible in Kotlin?) Or will the property values be made available via some other mechanism? If the latter, will a similar mechanism be added to Groovy?

Obviously, it would be nice to have this feature. Do you think it will take long to implement? Do you know that this can be implemented without necessarily causing problems elsewhere such that it can’t be implemented properly?

If it looks like it will take a long time to implement, will you only release the complementary change in Groovy when Kotlin has been similarly updated? I personally don’t see the need to hold back an improvement for Groovy if Kotlin lags behind due, as long as any discrepancies are documented, but maybe you might prefer DSL consistency for new features…

**Comment by netvl** [comment]
FWIW, since Kotlin allows using arbitrary expressions in the `version()` call, it is possible to define versions inside an object/class in the unnamed package in the `buildSrc` project, and then use that instead of project properties:

**buildSrc/src/main/kotlin/Versions.kt**:
```kotlin
object Versions {
    const val spotless = "3.24.0"
}
```

**build.gradle.kts**:
```kotlin
plugins {
    id("com.diffplug.gradle.spotless").version(Versions.spotless)
}
```

Unfortunately, the same is not possible in the Groovy DSL :(

**Comment by vlsi** [comment]
>I found this approach much cleaner than defining all versions in gradle.properties.

Versions in gradle.properties can be customized with startup arguments (e.g. `./gradlew -Pjacoco.version=...`) which is super helpful for integration testing.
If you bake the version into `Versions.kt`, then it is hard to update from the command line, and every version update would result in recompiling the build scripts.

**Comment by RuRo** [comment]
@netvl sorry to necro this thread. Can you please elaborate on how exactly did you achieve this?

I created the `buildSrc/src/main/kotlin/Versions.kt` file with the same contents as in your comment, but I am getting an `Unresolved reference: Versions` in `build.gradle.kts`.

**Comment by netvl** [comment]
@RuRo I'm not sure what goes wrong in your case, but it should be sufficient to have the `Versions.kt` file in the `buildSrc/src/main/kotlin` directory. Also you have to have a `build.gradle.kts`/`build.gradle` in `buildSrc`, to signify that it is a build source directory. It can have something like
```kotlin
plugins {
    `kotlin-dsl`
}
```
and nothing else.

@vlsi 

> it is hard to update from the command line

Well, it was never a use case for us. Frankly, I've never seen such thing being done anywhere, so I'm curious — is this needed to test compatibility of a project against different versions? It's just in my experience it is much more common for a project to depend on a particular specific version of a dependency, and there is simply no place for varying the version in this way.

**Comment by vlsi** [comment]
> is this needed to test compatibility of a project against different versions?

Yes, compatibility testing is one of the cases. It does not hurt if the version is externalized to a property file, however, if the version is baked in `kt` or `java`, then it becomes harder to augment.

> Frankly, I've never seen such thing being done anywhere,

Here's a discussion for JUnit5: https://github.com/junit-team/junit5/pull/2236

Here's a full-blown plugin that maintains versions in a property file: https://github.com/jmfayard/refreshVersions

Checkstyle project uses lots of external projects to check if a newer Checkstyle still works. For instance, [pgjdbc](https://github.com/pgjdbc/pgjdbc) allows to override Checkstyle version from the command line, so Checkstyle team knows when a Checkstyle fix breaks something for pgjdbc.

Here's the case for libraries: https://github.com/apache/calcite/blob/d2c859098acb3210ac5cad90c5ea28a3317f0386/.github/workflows/main.yml#L64-L74
First Calcite Avatica libaray is installed to the `mavenLocal` repository with `1.0.0-dev-master-SNAPSHOT` version, then `Calcite` is built and tested with `-Pcalcite.avatica.version=1.0.0-dev-master-SNAPSHOT`, so we know that avatica-produced poms are valid, and we use unusual version to ensure we don't accidentally test with a stale snapshot.

>and there is simply no place for varying the version in this way.

Well, if you declare version as `group:module:1.0`, then it means `1.0 or never`. That is a varying version, and it is quite common.

**Comment by netvl** [comment]
@vlsi thanks for the explanation, this does make sense. I guess I just happened never to work with such cases)

> Well, if you declare version as `group:module:1.0`, then it means `1.0 or never`. That is a varying version, and it is quite common.

And this one I'm not sure I understand. If you don't use wildcards, then the version is fixed, isn't it? At least in Maven-based systems. Well, I guess it's possible to say that "failure to resolve" is a "variant" of the version, but this seems like a singular case, and it's kind of assumed by default (well, I certainly did not consider it a "variant" when I wrote about it :))

**Comment by vlsi** [comment]
>And this one I'm not sure I understand. If you don't use wildcards, then the version is fixed, isn't it? At least in Maven-based systems.

Well, in Maven `group:module:1.0` declares an intention rather than an exact version. Maven can select another version if there are other declarations of the same artifact with different version.

Here's a quiz regarding "which version will be chosen": http://andresalmiray.com/maven-dependencies-pop-quiz-results/

Here's the quiz with similar questions on Gradle (it does provide a reference to the documentation that justifies Gradle behavior!): https://docs.google.com/forms/d/e/1FAIpQLScxNz27uKbdco2zaq2AkNiXzLRyTuRb1EO4MB7-0r7A06P2AQ/viewform

**Comment by netvl** [comment]
Well, automatic version upgrade due to conflict resolution does make the version kind of "dynamic", in the sense that it might not be exactly the same as declared, but as long as all artifacts in the dependency tree are immutable (so e.g. some `group:module:1.0` is not republished with a different POM), and as long as you don't use wildcard dependencies, then the final version set is still completely static and won't change ever without manual interference. So what I meant under "varying" is changing a particular dependency version for a particular build run, not variability of the version due to conflict resolution. But I now see how it could be useful, so thanks for the explanation)

**Comment by elect86** [comment]
Any news on this one?

**Comment by pumano** [comment]
2k21 here and NO SUPPORT for that feature!

**Comment by pumano** [comment]
2k21 here and NO SUPPORT here...

**Comment by jakzal** [comment]
There's a workaround to use plugin management: https://docs.gradle.org/current/userguide/plugins.html#sec:plugin_management

Example:

```gradle
// gradle.properties
kotlinVersion=1.5.10
```

```kotlin
// settings.gradle.kts
rootProject.name = "foo"

pluginManagement {
  val kotlinVersion: String by settings
  plugins {
    id("org.jetbrains.kotlin.jvm") version kotlinVersion
  }
}
```

```kotlin
// build.gradle.kts
plugins {
  id("org.jetbrains.kotlin.jvm")
}
```

This way I can either rely on the default kotlinVersion:

```
./gradlew test
```

Or override it from the command line:

```
./gradlew  clean test -PkotlinVersion=1.5.10
```

**Comment by sandwwraith** [comment]
Huge +1 to the issue. Workaround with `settings.gradle.kts` is a counterintuitive and hard to found on your own. My suggestion is to allow to write `val kotlinVersion: String by settings` (or `by properties`) directly in the `plugins {}` block of the main buildscript.

**Comment by jbartok** [comment]
Issue persists in latest Gradle versions.

**Comment by dkowis** [comment]
Since the last comment was over a year ago, I was just hunting this and would like to necro it as still being important.
